cli/src/main/java/org/adamalang/cli/router/CliElement.java

```diff
@@ -0,0 +1,169 @@
+package org.adamalang.cli.router;
+
+import org.adamalang.cli.runtime.ArgumentItem;
+import java.util.HashMap;
+import java.util.Map;
+
+public class CliElement {
+  public String name;
+  public String doc;
+  public Map<String, CliElement> Commands = new HashMap<>();
+  public Map<String, ArgumentItem> Arguments = new HashMap<>();
+  public static Map<String, ArgumentItem> fullArgumentList = populateArgs();
+  public static Map<String, CliElement> Groups = populateGroups();
+
+  public CliElement(HashMap<String, CliElement> commands, String name, String type, String doc) {
+    this.name = name;
+    this.doc = doc;
+    if (type.equals("group")) {
+      Commands = commands;
+    }
+  }
+
+  public CliElement(HashMap<String, ArgumentItem> arguments, String name, String doc) {
+    this.doc = doc;
+    this.name = name;
+    Arguments = arguments;
+  }
+  public static HashMap<String, CliElement> populateGroups() {
+    HashMap<String, CliElement> returnMap = new HashMap<>();
+
+    HashMap<String, ArgumentItem> initArgs = new HashMap<>();
+    returnMap.put("init", new CliElement(initArgs, "init", "Initializes the config with a valid token"));
+
+    HashMap<String, CliElement> spaceCommands = new HashMap<>();
+    HashMap<String, ArgumentItem> createSpaceArgs = new HashMap<>();
+    createSpaceArgs.put("--space", fullArgumentList.get("--space"));
+    spaceCommands.put("create", new CliElement(createSpaceArgs, "create", "Creates a new space"));
+
+    HashMap<String, ArgumentItem> deleteSpaceArgs = new HashMap<>();
+    deleteSpaceArgs.put("--space", fullArgumentList.get("--space"));
+    spaceCommands.put("delete", new CliElement(deleteSpaceArgs, "delete", "Deletes an empty space"));
+
+    HashMap<String, ArgumentItem> deploySpaceArgs = new HashMap<>();
+    deploySpaceArgs.put("--space", fullArgumentList.get("--space"));
+    deploySpaceArgs.put("--plan", fullArgumentList.get("--plan"));
+    deploySpaceArgs.put("--file", ArgumentItem.setOptionalFromMap(fullArgumentList, "--file", "null"));
+    spaceCommands.put("deploy", new CliElement(deploySpaceArgs, "deploy", "Deploy a plan to a space"));
+
+    HashMap<String, ArgumentItem> setRxhtmlSpaceArgs = new HashMap<>();
+    setRxhtmlSpaceArgs.put("--space", fullArgumentList.get("--space"));
+    setRxhtmlSpaceArgs.put("--file", ArgumentItem.setOptionalFromMap(fullArgumentList, "--file", "null"));
+    spaceCommands.put("set-rxhtml", new CliElement(setRxhtmlSpaceArgs, "set-rxhtml", "Set the frontend RxHTML forest"));
+
+    HashMap<String, ArgumentItem> getRxhtmlSpaceArgs = new HashMap<>();
+    getRxhtmlSpaceArgs.put("--space", fullArgumentList.get("--space"));
+    spaceCommands.put("get-rxhtml", new CliElement(getRxhtmlSpaceArgs, "get-rxhtml", "Set the frontend RxHTML forest"));
+
+    HashMap<String, ArgumentItem> uploadSpaceArgs = new HashMap<>();
+    uploadSpaceArgs.put("--space", fullArgumentList.get("--space"));
```

Comment:
From a design perspective, this could be considered "data embedding" where the code generator serves as a way to parse and validate data such that it goes into the code clean. From a manager hat, this works and solve the use-case.

However, from a craftsman perspective, it's a missed opportunity to really optimize and polish the screws.

In relationship to Argument which does all the heavy lifting from this data, a case could be made that the router could look something like this:
```
public static int route(String[] args, RootHandler handler, Output output) {
  if (args.length == 0) {
   displayRootHelp();
   return 1;
  }
  switch (args[0]) {
    case "space":
    case "spaces":
        SpaceHandler spaceHandler = handler.makeSpaceHandler();
        if (args.length == 1) {
          displaySpaceHelp();
          return 1;
        }
        switch (args[1]) {
            case "create": {
               SpaceCreateArgs args = SpaceCreateArgs.from(args, 2);
               if (args == null) {
                 SpaceCreateArgs.help();
                 return 1;
               }
               YesOrError out = output.makeYesOrError();
               spaceHandler.create(args, out);
               return out.result;
            }
        }
  }
```

The key idea is putting most of the validation and difficult work in the code generator, leaning up the runtime such that execution is exceptionally fast without map lookups.